crates/protocol/src/var_int.rs

```diff
@@ -0,0 +1,109 @@
+use std::{
+    io::{Cursor, Read, Write},
+    num::TryFromIntError,
+};
+
+use byteorder::ReadBytesExt;
+use thiserror::Error;
+
+use crate::io::{Decoder, Encoder};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct VarInt(pub i32);
+
+impl VarInt {
+    pub const MAX_SIZE: usize = 5;
+
+    pub const fn written_size(self) -> usize {
+        match self.0 {
+            0 => 1,
+            n => (31 - n.leading_zeros() as usize) / 7 + 1,
+        }
+    }
+
+    pub fn decode_partial(mut r: impl Read) -> Result<i32, VarIntDecodeError> {
```

Comment:
We should use `async` APIs everywhere: not only for consistency but also for the same handling of things like EOFs.

Reply:
This is the bit I'll be rewriting, but you should know that spamming awaits everywhere is probably going to slow the parser down. Sometimes it's better to build up a complete `Bytes` buffer in async (only jumping to Tokio code a few times and making minimal syscalls) and then read from it synchronously.

Reply:
That's a good point, but since:

* it is not a trait (thus boxed) `Future`, it is comparatively easy for the compiler to optimmize it (if not now, then in the future);
* there seems to be a single `await` point here.

I agree that this still may become a bottleneck, but as for now I would prefer a simpler, yet easier to use implementation and, if proven to be a bottlneck in bencmhmarks, optimize it in the future.